### DISCUSSION TITLE
Discussion: [Best approach for creating custom controls for WinUI]

### DISCUSSION THREAD

**Discussion by syncsiva** [microsoft/microsoft-ui-xaml#7493] [General]
# Discussion: [Best approach for creating custom controls for WinUI] 
What is the best approach for creating custom control? Is it good to develop controls in C# or C++. Any guidelines regarding the same will be useful?

## Related Links
https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/xaml-cust-ctrl#create-a-blank-app-bglabelcontrolapp
https://github.com/microsoft/microsoft-ui-xaml/issues/2863

**Comment by StephenLPeters** [comment]
It really depends on what you are trying to do. If you want the control to be usable by native components/apps you would probably want the control to be written in c++, otherwise loading your winrt component would require that the app load the CLR which is a non-starter for a lot of native apps.  You can see controls written in C++ in this repro, and controls written in c# in the [Windows Community Toolkit](https://github.com/windows-toolkit/WindowsCommunityToolkit).

**Comment by syncsiva** [comment]
@StephenLPeters  I'm working on creating scheduler control with day, week views similar outlook. I'm currently analyzing whether to start, using in Class Library (WinUI in UWP) or [C++/WinRT](https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/xaml-cust-ctrl). I want my custom control to work with WinUI for UWP and also WinUI for Desktop.

1. Whether developing the control in C++/WinRT will have performance benefits over developing custom control in Class Library (WinUI in UWP)? 
 2. If i develop control in C++/WinRT, how it can be consumed in WinUI in Desktop application? I'm new to WinUI.
